Started tests for gnm.

Endle opened the pull request:
Related to  #743

I followed commit 33e15b60df0a2a6e4e4a0b216e5125b7316c1eec , and added a test for gnm.



PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

Signed-off-by: Zhenbo Li <litimetal@gmail.com>

jbarnoud commented on testsuite/MDAnalysisTests/analysis/test_gnm.py:
```diff
@@ -0,0 +1,49 @@
+                'DCD parser not available. Are you using python 3?')
+    def setUp(self):
+        self.tmpdir = tempdir.TempDir()
+        self.universe = MDAnalysis.Universe(PSF, DCD)
```
Would the test make sense on `MDAnalysisTests.datafiles.XTC` rather than DCD? It would allow to run the test on python 3.

jbarnoud commented on testsuite/MDAnalysisTests/analysis/test_gnm.py:
```diff
@@ -0,0 +1,49 @@
+    def setUp(self):
+        self.tmpdir = tempdir.TempDir()
+        self.universe = MDAnalysis.Universe(PSF, DCD)
+        self.gnm = MDAnalysis.analysis.gnm.GNMAnalysis(self.universe)
```
This could go in the test itself rather than in the `setUp`. If the initialization of the analysis fails, it should be a failure of the test.

richardjgowers replied:
Is a failure in setup an Error or a Fail?

jbarnoud replied:
On 27/03/16 16:54, Richard Gowers wrote:
>
> In testsuite/MDAnalysisTests/analysis/test_gnm.py 
> <https://github.com/MDAnalysis/mdanalysis/pull/803#discussion_r57529750>:
>
> > +from numpy.testing import (TestCase, dec, assert_equal, assert_almost_equal)
> > +import numpy as np
> > +
> > +import tempdir
> > +
> > +from MDAnalysisTests.datafiles import PSF, DCD
> > +from MDAnalysisTests import parser_not_found
> > +
> > +
> > +class TestGNM(TestCase):
> > +    @dec.skipif(parser_not_found('DCD'),
> > +                'DCD parser not available. Are you using python 3?')
> > +    def setUp(self):
> > +        self.tmpdir = tempdir.TempDir()
> > +        self.universe = MDAnalysis.Universe(PSF, DCD)
> > +        self.gnm = MDAnalysis.analysis.gnm.GNMAnalysis(self.universe)
>
> Is a failure in setup an Error or a Fail?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/pull/803/files/08ce51ff1bddb8a0906d82801e0a3983229137d9#r57529750>
>
Since it is not going to raise an AssertError, it will be an error 
wherever is the line. That makes my argument above moot. But 
semantically, I still think creating the analysis object is part of the 
test.

Endle commented:
Thanks.
[In my new patch](https://github.com/Endle/mdanalysis/commit/6651bd071c79c91a8e594786d1149a9eb33e5c8e), I used GRO+XTC instead of PSF+DCD, which works fine on Python3.
Also, I moved `self.gnm = MDAnalysis.analysis.gnm.GNMAnalysis`

jbarnoud commented on testsuite/MDAnalysisTests/analysis/test_gnm.py:
```diff
@@ -0,0 +1,50 @@
+        del self.tmpdir
+
+    def test_gnm(self):
+        self.gnm = MDAnalysis.analysis.gnm.GNMAnalysis(self.universe)
```
Does not need to be in the instance namespace: `self.gnm` could be just `gnm`. Then it is not needed to delete it latter as it will just not live out of the function scope.

Endle commented:
> Does not need to be in the instance namespace: self.gnm could be just gnm. Then it is not needed to delete it latter as it will just not live out of the function scope.

Thank you for mentioning it.

jbarnoud commented:
Your test [covers about 60%](https://coveralls.io/jobs/13091370/source_files/111008912) of the gnm module. Do you want to tackle what is left? Or at least increase the coverage? As it is now, it is a good start but the module cannot be considered supported yet.

Otherwise, it looks good to me.

Endle commented:
Thanks.
I'm going to add tests for generate_kirchoff and backup_file
I'll keep contributing these tests to help me understand the code, and make my porting job easier (if my proposal is accepted)

Will you merge this PR, or wait for me to write a (mostly) completed test for gnm, and to send a new PR?

jbarnoud commented:
On 28/03/16 13:20, Zhenbo Li wrote:
>
> Thanks.
> I'm going to add tests for generate_kirchoff and backup_file
> I'll keep contributing these tests to help me understand the code, and 
> make my porting job easier (if my proposal is accepted)
>
> Will you merge this PR, or wait for me to write a (mostly) completed 
> test for gnm, and to send a new PR?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/pull/803#issuecomment-202344633>
>
If you continue to work on the same tests, I will wait. Add "[WIP]" at 
the beginning of the PR title to avoid that the PR get merged by accident.

Endle commented:
I added some tests, and [coverage reached 80%](https://coveralls.io/builds/5581771/source?filename=miniconda%2Fenvs%2Fpyenv%2Flib%2Fpython2.7%2Fsite-packages%2FMDAnalysis%2Fanalysis%2Fgnm.py), leaving `backup_file`, which is not needed to be tested.

I added tests for closeContactGNMAnalysis, but such test is very slow. It costed 294.926s on my computer(before adding it, it costs me only 15.524s). So I added `@attr('slow')` for it.

Can this PR be merged now?

jbarnoud commented:
Great! This seems mergeable. There is a few cases you miss and that do not seem difficult to test:

* `GNMAnalysis.run` with `skip != 1`
* `GNMAnalysis` with `ReportVector` set
* `closeContactGNMAnalysis` with `MassWeigt == False`

Are you willing to give them a try?

Endle commented:
Tests for these situations added.
